### PR TITLE
feat(db): Don't store `TxHash` when writing to the table

### PR DESCRIPTION
**Pull request by joshieDo**
* Don't store `TxHash` on the database.
* Add new type for DB insertion: `TransactionSignedNoHash`

This should highly decrease the storage space used by `Transactions` table. However `tx-lookup` is greatly slowed down on its current form. We can improve it with rayon, but it will still be slower overall, since it has to calculate the hash for each tx. 

(random thought: wondering if we could merge this stage for when we first calculate the tx hash?)

**2M blocks:**
```
main: 1.6G
rlp: 1.3G 
compact - 1.3G
```

**5M blocks**
```
main: 30.6 G
rlp: 26.1 G
compact: 25.4 G
```

**2M blocks** `(dbmain - main branch | dbb2 - this branch)`

image

While changing `senders_recovery` noticed a potential improvement (~10%)

### CODE REVIEW

**Comment by gakonst** [comment]
Nice! What other areas are you looking at around the DB/Codecs to improve on size & i/o?

**Comment by joshieDo** [comment]
Updated the original PR description.

It would be cool if @mattsse  could take a look at this to make sure we're not calculating the tx hash during the network/pool too many times...

**Comment by gakonst** [comment]
> (random thought: wondering if we could merge this stage for when we first calculate the tx hash?)

Can you unpack this a bit?

**Comment by gakonst** [comment]
> Updated the original PR description.
> 
> 
> 
> It would be cool if @mattsse  could take a look at this to make sure we're not calculating the tx hash during the network/pool too many times...

@mattsse fyi

**Comment by joshieDo** [comment]
Adding rayon to calculate the tx hash of all txes, made it slightly faster (38 -> 32).

> Can you unpack this a bit?

During the `Body` stage, if we already (i'm not sure if we do) have the transaction hash, maybe we could just insert it into the TxLookup table there? @onbjerg

**Review by mattsse — CHANGES_REQUESTED** [review]
> Don't store TxHash on the database.

I see we're doing this here

https://github.com/paradigmxyz/reth/blob/a22f9880970fbacf8d21c478b8bcd5a28a00701f/crates/storage/db/src/tables/mod.rs#L171-L174

wouldn't we get the same result if we'd change the type here

if this is only for the database, should we instead add a new type for DB inserts only?

or are there additional benefits?

File: crates/primitives/src/transaction/mod.rs
Comment (joshieDo) [reply crates/primitives/src/transaction/mod.rs:695]:
```rust
self.hash.get_or_insert_with(|| self.recalculate_hash());
```
this cannot work though, since `self.hash` is a mutable borrow, which is followed by a disallowed immutable one `|| self.recalculate_hash`

**Comment by onbjerg** [comment]
> During the `Bodies` stage, if we already (i'm not sure if we do) have the transaction hash, maybe we could just insert it into the TxLookup table there? @onbjerg

The downloader gives us `TransactionSigned`, but I am not sure where exactly the hash is computed as it is not transmitted over the network. It seems like it is calculated somewhere in the networking component? cc @mattsse 

We could probably insert it in the bodies downloader w/o much issue.

File: crates/primitives/src/transaction/mod.rs
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:667]: I'd like a one-liner why we have this (used for DB)

File: crates/primitives/src/transaction/mod.rs
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:766]: I see, this is now a util function to get hash ref.
This is fine.

**Comment by joshieDo** [comment]
> if this is only for the database, should we instead add a new type for DB inserts only?
I just thought that we had too many Transaction types... But yeah, this way is cleaner. Updating the PR description

Also noticed something on `senders_recovery` which could be improved, and resulted in a 10% performance improvement.

image

File: crates/primitives/src/transaction/mod.rs
Comment (mattsse) [inline crates/primitives/src/transaction/mod.rs:721]: I'd also like a function that does this (`with_hash()` perhaps?)
The reason this is more convenient is because we now have a few layers, imagine you need to go from `TransactionSignedNoHash` to `TransactionSignedRecovered` with only `From` impls.

with dedicated functions you can do `no_hash.with_hash().into_erecovered()`

**Review by mattsse — CHANGES_REQUESTED** [review]
sweet, this is a pretty grood improvement re disk space.

only a few nits

**Comment by mattsse** [comment]
> I just thought that we had too many Transaction types...

Yes, unfortunately, there is no way around it, but one more reason to use very specific types.

**Comment by joshieDo** [comment]
i think in this case, the iai test failure can be ignored. There are more RAM accesses on decompress (and probably just fluctuation)  , but overall less instructions

**Review by mattsse — APPROVED** [review]
lgtm,

defer to @onbjerg when to merge this so it doesn't cause issues with other pending prs

File: crates/stages/src/stages/sender_recovery.rs
Comment (gakonst) [inline crates/stages/src/stages/sender_recovery.rs:128]: smart!

**Review by gakonst — APPROVED** [review]

**Comment by gakonst** [comment]
@joshieDo this is good to merge after resolving the conflicts. we merged block level indexing so best to turbo merge this so we avoid too many db breakages back to back in our testing

**Review by mattsse — APPROVED** [review]

**Pull request merged**
